File: Codeforces/75C - Modified GCD.py
# ==================================================
# Problem  :   75C - Modified GCD
# Run time :   0.186 sec.
# Language :   Python 3.5
# ==================================================


def gcd(m, n):
	while n != 0:
		t = m % n
		m = n
		n = t

	return m


def get_divisors(g):
	from math import sqrt
	
	rt = int(sqrt(g))
	divisors = []

	for x in range(1, rt+1):
		if g % x == 0:
			divisors.append(x)
			divisors.append(g // x)

	if divisors[-1] ==  divisors[-2]:
		divisors.pop()

	divisors.sort()

	return divisors


def main():
	from sys import stdin, stdout
	from bisect import bisect_right

	# stdin = open("input.txt", "r")

	it = iter(map(int, stdin.read().split()))

	a, b = next(it), next(it)
	divisors = get_divisors(gcd(a, b))

	n = next(it)
	ans = []

	for _ in range(n):
		low, high = next(it), next(it)
		
		idx = bisect_right(divisors, high)
		d = divisors[idx-1]

		ans.append(str(d if low <= d <= high else -1))
		ans.append('\n')

	stdout.write(''.join(ans))


if __name__ == '__main__':
	main()

File: Codeforces/75C - Modified GCD.cpp
// ==================================================
// Problem  :   75C - Modified GCD
// Run time :   0.062 sec.
// Language :   C++11
// ===================================================

#include <cstdio>
#include <algorithm>
#include <vector>
#include <cmath>
using namespace std;

vector<int> divisors;

int gcd(int m, int n)
{
    while(n != 0) {
        int t = m % n;
        m = n;
        n = t;
	}

    return m;
}

void get_divisors(int g)
{
    int rt = sqrt(g);

    for(int i = 1; i <= rt; ++i) {
        if(g % i == 0) {
            divisors.push_back(i);
            divisors.push_back(g/i);
        }
    }

    if(divisors.back() == rt) divisors.pop_back();  // rt appears twice.

    sort(divisors.begin(), divisors.end());
}

int main()
{
    //freopen("in", "r", stdin);

    int a, b;
    scanf("%d %d", &a, &b);

    get_divisors(gcd(a, b));

    int n;
    scanf("%d", &n);

    int low, high;

    while(n--) {
        scanf("%d %d", &low, &high);

        int idx = upper_bound(divisors.begin(), divisors.end(), high) - divisors.begin();

        printf("%d\n", ((idx > 0 && low <= divisors[idx-1])? divisors[idx-1] : -1));
    }

    return 0;
}
